### PR TITLE
1.2/968 write a voucher removal function so recorded ed vouchers can be rolled back include this in voucher controller

### PR DESCRIPTION
**Pull request by charlesstrange2**
Right, lets try this again.

System should now try to pick the correct reverse transition back to the original state based on the voucher *prior* state. Also, added more sensible messages in translations, and updated tests to cope.

### CODE REVIEW

File: app/Http/Controllers/API/VoucherController.php [hunk 1]
```diff
@@ -181,42 +201,56 @@ private function constructResponseMessage($responses)
...
                     break;
             }
         } else {
             return [
                 // Todo: This message needs work - but not this round.
-                'message' => trans('api.messages.batch_voucher_submit', [
+                'message' => trans(
+                    'api.messages.batch_voucher_submit',
+                    [
                     'success_amount' => count($responses['success']),
```
Comment (katjam) [inline app/Http/Controllers/API/VoucherController.php]: Does this need to be `success_add` and `success_reject`. I don't think the response has `success` key anymore.

File: app/Listeners/StateHistoryManager.php [hunk 2]
```diff
@@ -33,7 +33,7 @@ public function postTransition(TransitionEvent $event)
 
         $model->history()->create([
             "transition" => $event->getTransition(),
-            "from" => "", // what the state was before.
+            "from" => $event->getState(), // what the state was before.
```
Comment (katjam) [inline app/Listeners/StateHistoryManager.php]: Will it be empty if there wasn't one? I think there was a trello card about this way back when - to implement when we started to care that the `from`s were not being recorded. I guess you can locate and annihilate?

File: app/Traits/Statable.php [hunk 3]
```diff
@@ -36,6 +36,16 @@ public function getStateMachine()
...
     }
 
+    /**
+     * gets the prior state associated with the Stateable model.
+     *
+     * @return self::HISTORY_MODEL
+     */
+    public function getPriorState()
+    {
+        return $this->history()->get(null)->last();
```
Comment (katjam) [inline app/Traits/Statable.php]: Cool is the output predictable?

File: config/state-machine.php [hunk 4]
```diff
@@ -50,6 +50,18 @@
                 'from' => ['printed','dispatched','allocated'],
                 'to' =>  'recorded',
             ],
+            'reject-to-printed' => [
```
Comment (katjam) [inline config/state-machine.php]: So complicated! :)

File: app/Http/Controllers/API/VoucherController.php [hunk 5]
```diff
@@ -115,10 +133,12 @@ public function transition(Request $request)
         // Currently we can
         //  'collect' - submit and
         //  'confirm' - request payment on.
-        $responses['success'] = $success_codes;
+        $responses['success_add'] = $success_codes;
+        $responses['success_reject'] = $reject_codes;
         $responses['own_duplicate'] = $own_duplicate_codes;
         $responses['other_duplicate'] = $other_duplicate_codes;
         $responses['invalid'] = $bad_codes;
+        $responses['failed_reject'] = $failed_rejects;
```
[the remark this answers is not in the document]
Comment (charlesstrange2) [reply app/Http/Controllers/API/VoucherController.php]: No, it;s an aweful pile of arrays that hold the exception states for each voucher transition.
Ideally i'd OO the whole thing - but time is short.

File: app/Http/Controllers/API/VoucherController.php [hunk 6]
```diff
@@ -181,42 +201,56 @@ private function constructResponseMessage($responses)
...
         if ($total_submitted === 1) {
             switch ($error_type) {
-                case 'success':
+                case 'success_add':
                     return [
-                        'message' => trans('api.messages.voucher_success')
+                        'message' => trans('api.messages.voucher_success_add'),
                     ];
                     break;
+                case 'success_reject':
```
[the remark this answers is not in the document]
Comment (charlesstrange2) [reply app/Http/Controllers/API/VoucherController.php]: when a trader wants to reject a voucher inserted by error, yes.

File: app/Http/Controllers/API/VoucherController.php (hunk 1 above)
Comment (charlesstrange2) [reply app/Http/Controllers/API/VoucherController.php]: Hmm. will check that...

File: app/Listeners/StateHistoryManager.php (hunk 2 above)
Comment (charlesstrange2) [reply app/Listeners/StateHistoryManager.php]: yes, I think it pops out null.

File: app/Traits/Statable.php (hunk 3 above)
Comment (charlesstrange2) [reply app/Traits/Statable.php]: yes, unlike `latest()` which can't tell between vouchers `created_at` in the same second, fetching the history orders by ID for `last()` to pop the end off. It shouldn't be a memory problem, in theory, as there shouldn't be very many transitions.

File: config/state-machine.php (hunk 4 above)
Comment (charlesstrange2) [reply config/state-machine.php]: I know. :cry: 

I thought "revert", but it isn't reverting, really. that implies we rollback the whole transaction and remove it, whereas i think we need a record of all transactions.

**Pull request merged**